src/test/java/com/karankumar/bookproject/ui/shelf/CustomShelfFormTest.java

```diff
@@ -0,0 +1,169 @@
+
+    private CustomShelfForm customShelfForm;
+
+    @BeforeAll
+    public static void discoverRoutes() {
+        routes = new Routes().autoDiscoverViews("com.karankumar.bookproject.ui");
+    }
+
+    @BeforeEach
+    public void setup(@Autowired BookService bookService,
+                      @Autowired PredefinedShelfService predefinedShelfService,
+                      @Autowired CustomShelfService customShelfService) {
+        final SpringServlet servlet = new MockSpringServlet(routes, ctx);
+        MockVaadin.setup(UI::new, servlet);
+
+        customShelfForm = new CustomShelfForm(customShelfService, predefinedShelfService);
+
+    }
+
+    @Test
+    void saveButtonShouldBeInitiallyGreyedOut () {
+        // given
+        // customShelfForm
+
+        // when
+        userOpenDialog();
+
+        // then
+        formIsInInitialState();
+    }
+
+    private void userOpenDialog() {
+        customShelfForm.addShelf();
+    }
+
+    private void formIsInInitialState() {
+        Assertions.assertTrue(customShelfForm.shelfNameField.getValue().equals(""));
+        Assertions.assertFalse(customShelfForm.shelfNameField.isInvalid());
+        Assertions.assertFalse(customShelfForm.saveButton.isEnabled());
+    }
+
+    @Test
+    void saveButtonShouldBeGreyedOutWithExistingShelfName () {
+        // given
+        // customShelfForm
+
+        // when
+        shelfNameIsLikeOneAlreadyInUse();
+
+        // then
+        formsIsInErrorState();
+    }
+
+    private void shelfNameIsLikeOneAlreadyInUse() {
+        customShelfForm.shelfNameField.setValue(PredefinedShelf.ShelfName.TO_READ.toString());
+    }
+
+    private void formsIsInErrorState() {
+        Assertions.assertTrue(customShelfForm.shelfNameField.isInvalid());
+        Assertions.assertFalse(customShelfForm.saveButton.isEnabled());
+    }
+
+    @Test
+    void saveButtonShouldBeEnabledWithNonExistingShelfName () {
+        // given
+        // customShelfForm
+
+        // when
+        shelfNameIsNotLikeOneAlreadyInUse();
+
+        // then
+        formIsInValidState();
+    }
+
+    private void shelfNameIsNotLikeOneAlreadyInUse() {
+        customShelfForm.shelfNameField.setValue("NonExistingShelfName");
+    }
+
+    private void formIsInValidState() {
+        Assertions.assertNotEquals("", customShelfForm.shelfNameField.getValue());
+        Assertions.assertTrue(customShelfForm.saveButton.isEnabled());
+    }
+
+    @Test
+    void savingAndReopeningFormShouldClearTextfield () {
+        // given
+        // customShelfForm
+
+        // when
+        userOpenDialog();
+        userCloseSavingAShelf();
+        userOpenDialog();
+        // then
+        formIsInInitialState();
+    }
+
+    private void userCloseSavingAShelf() {
+        customShelfForm.shelfNameField.setValue("Test");
+        customShelfForm.saveButton.click();
+    }
+
+    @Test
+    void closingWithoutSavingAndReopeningFormShouldClearTextfield () {
```

Comment:
Could you make the 'f' in 'field' uppercase?

Reply:
Sure!